lib/src/git.rs

```diff
@@ -348,7 +366,38 @@ pub fn export_refs(
             failed_branches.push(branch_name);
         }
     }
-    Ok(failed_branches)
+    Ok((GitRefView::default(), failed_branches))
+}
+
+/// Reads the last seen state of git refs from a file, updates it with the
+/// provided function, and writes the result back to disk.
+///
+/// The state currently tracks only the local branches.
+fn with_last_seen_refs<Ret, Err, Err2: From<GitRefViewError> + From<Err>>(
+    repo_path: PathBuf,
+    f: impl FnOnce(&GitRefView) -> Result<(GitRefView, Ret), Err>,
+) -> Result<Ret, Err2> {
+    // Short-term TODO 1: Modification of this file, as well as the git refs in the
+    // git storage, should be protected by a lock.
+    //
+    // TODO 2: Conceptually, this state should contain all the information in the
+    // repo that's not stored by git and should not be affected by `jj undo`. In
+    // particular, jj's last seen view of remote-tracking branches should be stored
+    // here. It is currently affected by `undo`, which can cause minor issues.
```

Comment:
What I meant to do here is to contrast the contents of commits -- for which the git repo is authoritative and the only place they are stored -- and the "last seen" git state (which not literally stored by git, as it is different from the current git state)

I'm thinking about how to clarify this comment. Perhaps we should say something about storing information that can be potentially *conflicted* between jj and git.

Reply:
Maybe something like this:

```
Conceptually, this state should contain all the information in the
repo that can be updated by either git or jj. For example, branches should
be stored here, but remote-tracking branches and other git refs should not.
```

Actually, what problem does "It is currently affected by `undo`, which can cause minor issues." refer to?

Reply:
>  what problem does "It is currently affected by undo, which can cause minor issues." refer to?

Mainly, `jj undo` after `jj git fetch/push` can move remote tracking branches. If we use the remote tracking branches for `--force-with-lease`, it could become a bigger problem.

An example that I mean to turn into a test (unless I already have):

```
jj git push --branch a
jj branch s a -r somewhere
jj git push --branch a
jj op restore @--
```

At this point, `jj git push` *should* push the branch `a`, but I think it won't because it will consider the branch unchanged.

Reply:
> At this point, another jj git push --branch a should push the branch a, but I think it won't because it will consider the branch unchanged.

IIRC, `git push` doesn't check the 'local' value of the remote branch (refs/remotes/my-remote/my-branch on the local clone); it only checks the 'remote' value of the remote branch (refs/heads/my-branch on the remote). So if `jj git push` replicates `git push` closely (that depends on libgit2 and whether we added extra steps), it should push the branch.